### DISCUSSION TITLE
Where do I find the .whl file to install it?

### DISCUSSION THREAD

**Discussion by Centurax** [facebookresearch/xformers#717] [Q&A] [answered]
Hi I am kind of new to all this and basically this feels like a stupid question.

To give you context I am attempting to get Dreambooth working with Stable Diffusion and I get this:

> Your version of xformers is 0.0.16rc425.
> xformers >= 0.0.17.dev is required to be available on the Dreambooth tab.
> Torch 1 wheels of xformers >= 0.0.17.dev are no longer available on PyPI,
> but you can manually download them by going to:
> https://github.com/facebookresearch/xformers/actions
> Click on the most recent action tagged with a release (middle column).
> Select a download based on your environment.
> Unzip your download
> Activate your venv and install the wheel: (from A1111 project root)
> cd venv/Scripts
> activate
> pip install {REPLACE WITH PATH TO YOUR UNZIPPED .whl file}
> Then restart your project.

I tried using the setup file and got this:

> (venv) ...\stable-diffusion-webui\venv\Scripts> pip install ...\xformers-main\setup.whl
> WARNING: Ignoring invalid distribution -rotobuf (...stable-diffusion-webui\venv\lib\site-packages)
> WARNING: Requirement '...\stable-diffusion-webui\\scripts\\xformers-main\\setup.whl' looks like a filename, but the file does not exist
> ERROR: setup.whl is not a valid wheel filename.
> WARNING: Ignoring invalid distribution -rotobuf (...\stable-diffusion-webui\venv\lib\site-packages)
> WARNING: Ignoring invalid distribution -rotobuf (...\stable-diffusion-webui\venv\lib\site-packages)
> WARNING: Ignoring invalid distribution -rotobuf (...\stable-diffusion-webui\venv\lib\site-packages)

here is the question where is the .whl file to install it?  

Also are there are any steps I am unaware of to get this working?

**Comment by danthe3rd** [comment]
Hi,
You can go there:
https://github.com/facebookresearch/xformers/actions/workflows/wheels.yml

Then select the latest completed job (for example [this one](https://github.com/facebookresearch/xformers/actions/runs/4598589407)), then download the zip file that corresponds to your setup (for instance [ubuntu-22.04-py3.10-torch1.13.1+cu117](https://github.com/facebookresearch/xformers/suites/11995312356/artifacts/630137642))

**Reply by Centurax** [reply]
Hi,

Thanks for the reply

Unfortunately I am still a little lost with all this, I have a better idea what to look for now, however, I still have no idea where to download it from.  

I probably should have added I was looking for a Windows build when I wrote the first post.

**Reply by danthe3rd** [reply]
What is your python version (`python --version`) and your pytorch version (`python -c "import torch; print(torch.__version__)"`)?

**Reply by Centurax** [reply]
Python Version = Python 3.10.6

pytorch version just came back with:

> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
> ModuleNotFoundError: No module named 'torch'

**Reply by danthe3rd** [reply]
I believe you should reach out to whoever wrote this tutorial you are linking in the original post, as the rest will depend on your pytorch version, and I'm not sure which one you should use for that project

**Reply by sivert-io** [reply, marked as answer]
> Unfortunately I am still a little lost with all this, I have a better idea what to look for now, however, I still have no idea where to download it from.

### 1. Go here
https://github.com/facebookresearch/xformers/actions/workflows/wheels.yml

### 2. Click the top entry (in this case `ptxas: Build with O2 instead of O3`):
![image](https://user-images.githubusercontent.com/6785315/234533780-660c425f-e795-4419-af75-776fc294afd7.png)

### 3. Scroll down and click the correct version according to your project (most likely `windows-2019-py3.10-torch1.13.1`)
![image](https://user-images.githubusercontent.com/6785315/234533980-8324c757-e13b-423c-9ac3-46894b8b0bff.png)

### 4. Open the zip and extract the .whl file to the root of your project (in my case `H:\automatic1111`)
![image](https://user-images.githubusercontent.com/6785315/234534657-df8d3555-157c-4c16-b21a-3a462f2987a5.png)

### 5. Open a terminal (or cmd) in the root of your Stable Diffusion folder (in my case `H:\automatic1111`) and do:
```sh
cd venv/scripts
./activate
```

### 6. CD back to your root folder:
```sh
cd ..
cd ..
```

### 7. Install whl:
```sh
pip install xformers-0.0.19.dev523-cp310-cp310-win_amd64.whl
```

### 8. Start WEB-UI and look for this message:
`[+] xformers version 0.0.19.dev523 installed.`

### Example startup
![image](https://user-images.githubusercontent.com/6785315/234535342-abde6553-32e5-467c-b9d2-ebf329173849.png)

# Happy generating 😃

**Reply by Centurax** [reply]
thanks :)

**Reply by Pertenyos** [reply]
Hi, it does't seem to work for me. I do everything as suggested but it keeps telling me no xformers.

**Reply by curiouspers** [reply]
I was also fighting this, but on latest versions all you need to do is start automatic1111 with argument --xformers
It will install everything that is needed

**Reply by sivert-io** [reply]
> Hi, it does't seem to work for me. I do everything as suggested but it keeps telling me no xformers.

Did you add the --xformers argument in the webui-user.bat file?

**Reply by LineKernel** [reply]
you can cmd to your a1111 root folder then type webui-user.bat --xformers to start it once with xformers , it doesnt work for me tho (windows) 
i tried the marked answered but it broke my torch (torch cant use gpu anymore , even when i bypass the cuda test)
i m reinstalling right now

**Reply by LineKernel** [reply]
adding --xformers to webui-user.bat worked for me finally

**Reply by sgtsixpack** [reply]
`set COMMANDLINE_ARGS= --opt-sub-quad-attention --medvram --autolaunch
The syntax of the command is incorrect.
Already up to date.
venv "J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\Scripts\Python.exe"
fatal: No names found, cannot describe anything.
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Version: <none>
Commit hash: ba780a8e4fb27d1eb9595947380a63bd84fc419e
Installing requirements


Warning: caught exception 'Torch not compiled with CUDA enabled', memory monitor disabled
If submitting an issue on github, please provide the full startup log for debugging purposes.

Initializing Dreambooth
Dreambooth revision: dc413a14379b165355502d9f65856c40a4bb5b6f
Successfully installed accelerate-0.19.0 fastapi-0.94.1 gitpython-3.1.31 transformers-4.29.2


Does your project take forever to startup?
Repetitive dependency installation may be the reason.
Automatic1111's base project sets strict requirements on outdated dependencies.
If an extension is using a newer version, the dependency is uninstalled and reinstalled twice every startup.

[+] xformers version 0.0.20 installed.
[+] torch version 2.0.0 installed.
[+] torchvision version 0.15.1 installed.
[+] accelerate version 0.19.0 installed.
[+] diffusers version 0.16.1 installed.
[+] transformers version 4.29.2 installed.
[+] bitsandbytes version 0.35.4 installed.

Launching Web UI with arguments: --opt-sub-quad-attention --xformers --autolaunch
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: -low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: .low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 0low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 1low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 2low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 3low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 4low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 5low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 6low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 7low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 8low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: 9low is an invalid version and will not be supported in a future release
  warnings.warn(
J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\pkg_resources\__init__.py:123: PkgResourcesDeprecationWarning: llow is an invalid version and will not be supported in a future release
  warnings.warn(
WARNING[XFORMERS]: xFormers can't load C++/CUDA extensions. xFormers was built for:
    PyTorch 2.0.1+cu118 with CUDA 1108 (you have 2.0.0+cpu)
    Python  3.10.11 (you have 3.10.6)
  Please reinstall xformers (see https://github.com/facebookresearch/xformers#installing-xformers)
  Memory-efficient attention, SwiGLU, sparse and more won't be available.
  Set XFORMERS_MORE_DETAILS=1 for more details
Warning: caught exception 'Torch not compiled with CUDA enabled', memory monitor disabled
2023-06-10 11:30:21,563 - ControlNet - INFO - ControlNet v1.1.220
ControlNet preprocessor location: J:\AI training\Stable diffusion\stable-diffusion-webui-directml\extensions\sd-webui-controlnet\annotator\downloads
2023-06-10 11:30:21,628 - ControlNet - INFO - ControlNet v1.1.220
Loading weights [b0d8ba7265] from J:\AI training\Stable diffusion\stable-diffusion-webui-directml\models\Stable-diffusion\janaDefi_v24.safetensors
Creating model from config: J:\AI training\Stable diffusion\stable-diffusion-webui-directml\models\Stable-diffusion\janaDefi_v24.yaml
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
Textual inversion embeddings loaded(4): bad-hands-5, bad-image-v2-39000, easynegative, opulent beauty
Model loaded in 3.6s (load weights from disk: 0.1s, create model: 0.3s, apply weights to model: 0.6s, apply half(): 0.6s, move model to device: 1.9s).
Applying optimization: sub-quadratic... done.
CUDA SETUP: Loading binary J:\AI training\Stable diffusion\stable-diffusion-webui-directml\venv\lib\site-packages\bitsandbytes\libbitsandbytes_cudaall.dll...
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 9.6s (import torch: 1.4s, import gradio: 0.9s, import ldm: 0.4s, other imports: 0.7s, load scripts: 5.4s, create ui: 0.4s, gradio launch: 0.2s).
`
This is my webui-user.bat:
@echo off

set PYTHON=
set GIT=
set VENV_DIR=
@echo set COMMANDLINE_ARGS= --opt-sub-quad-attention --medvram --autolaunch
set /export REQS_FILE=.\extensions\sd_dreambooth_extension\requirements.txt
set COMMANDLINE_ARGS= --opt-sub-quad-attention --xformers --autolaunch
git pull

call webui.bat

I  have an AMD 6800xt so no cuda.  I'm trying to train with dreambooth.  Any suggestions?  Thx.

**Comment by coleea** [comment]
Check this
https://download.pytorch.org/whl/xformers/
